presto-spi/src/main/java/io/prestosql/spi/connector/ConnectorMetadata.java

```diff
@@ -223,6 +226,14 @@ default ColumnMetadata getColumnMetadata(ConnectorSession session, ConnectorTabl
         return emptyMap();
     }
 
+    /**
+     * Returns a (table name, Optional.empty()) entry if the table is redirected. See {@link #redirectTable}'s doc for details.
+     */
+    default Map<SchemaTableName, Optional<List<ColumnMetadata>>> listTableColumnsWithRedirection(ConnectorSession session, SchemaTablePrefix prefix)
+    {
+        return listTableColumns(session, prefix).entrySet().stream().collect(toUnmodifiableMap(Map.Entry::getKey, entry -> Optional.of(entry.getValue())));
```

Comment:
Shouldn't the default implementation actually call `redirectTable`?

Something like
```
return listTableColumns(session, prefix).entrySet().stream()
        .map(entry -> {
            if (redirectTable(session, entry.getKey()).isPresent()) {
                return new SimpleEntry<SchemaTableName, Optional<List<ColumnMetadata>>>(entry.getKey(), Optional.empty());
            }
            return new SimpleEntry<>(entry.getKey(), Optional.of(entry.getValue()));
        })
        .collect(toUnmodifiableMap(Entry::getKey, Entry::getValue));
```

Reply:
This makes sense. But I think there's a trade-off here:
1. Let the default implementation call `redirectTable`. So it's fine even if a connector implements `redirectTable` and not `listTableColumnsWithRedirection`. However, this makes other connectors less efficient, e.g. those which don't implement table redirection APIs
2. Don't let the default implementation call `redirectTable` and make connector authors responsible for also implementing `listTableColumnsWithRedirection` when implementing `redirectTable`. The pro and con are the contrary.

What do you folks think? @findepi @electrum

Reply:
> However, this makes other connectors less efficient, e.g. those which don't implement table redirection APIs

isn't it that the difference will be negligible? am i missing something?

Reply:
@lxynov was this covered?